### PR TITLE
[Relay] No unification of arg type and func type

### PR DESCRIPTION
**Pull request by icemelon9**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

File: src/relay/pass/type_solver.cc [hunk 1]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+    }
+
+    auto ft1 = GetRef<FuncType>(op);
+    auto ft2 = Downcast<FuncType>(Bind(GetRef<FuncType>(ftn), subst_map));
+
+    Type ret_type = Join(ft1->ret_type, ft2->ret_type);
+
+    std::vector<Type> arg_types;
+    for (size_t i = 0; i < ft1->arg_types.size(); i++) {
+      Type arg_type = Join(ft1->arg_types[i], ft2->arg_types[i]);
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: This should be meet. function argument should be contravariant.

File: src/relay/pass/type_solver.cc [hunk 2]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+    return RefTypeNode::make(Join(op->value, rtn->value));
+  }
+
+  Type VisitType_(const TypeCallNode* op, const Type& tn) override {
+    const auto* tcn = tn.as<TypeCallNode>();
+    if (!tcn || tcn->args.size() != op->args.size()) {
+      return Type();
+    }
+
+    Type func = Join(op->func, tcn->func);
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: TypeCallNode should be invariant.

File: src/relay/pass/type_solver.cc [hunk 3]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+
+    return FuncTypeNode::make(arg_types, ret_type, ft1->type_params, type_constraints);
+  }
+
+  Type VisitType_(const RefTypeNode* op, const Type& tn) final {
+    const auto* rtn = tn.as<RefTypeNode>();
+    if (!rtn) {
+      return Type(nullptr);
+    }
+    return RefTypeNode::make(Join(op->value, rtn->value));
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: Ref should be invariant.

File: src/relay/pass/type_solver.cc [hunk 4]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+    }
+
+    auto ft1 = GetRef<FuncType>(op);
+    auto ft2 = Downcast<FuncType>(Bind(GetRef<FuncType>(ftn), subst_map));
+
+    Type ret_type = Meet(ft1->ret_type, ft2->ret_type);
+
+    std::vector<Type> arg_types;
+    for (size_t i = 0; i < ft1->arg_types.size(); i++) {
+      Type arg_type = Meet(ft1->arg_types[i], ft2->arg_types[i]);
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: should be join

File: src/relay/pass/type_solver.cc [hunk 5]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+
+    return FuncTypeNode::make(arg_types, ret_type, ft1->type_params, type_constraints);
+  }
+
+  Type VisitType_(const RefTypeNode* op, const Type& tn) final {
+    const auto* rtn = tn.as<RefTypeNode>();
+    if (!rtn) {
+      return Type(nullptr);
+    }
+    return RefTypeNode::make(Meet(op->value, rtn->value));
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: like above comment, should be invariant

File: src/relay/pass/type_solver.cc [hunk 6]
```diff
@@ -90,11 +96,511 @@ class TypeSolver::OccursChecker : public TypeVisitor {
...
+  Type VisitType_(const TypeCallNode* op, const Type& tn) override {
+    const auto* tcn = tn.as<TypeCallNode>();
+    if (!tcn || tcn->args.size() != op->args.size()) {
+      return Type();
+    }
+
+    Type func = Meet(op->func, tcn->func);
+    tvm::Array<Type> args;
+    for (size_t i = 0; i < op->args.size(); i++) {
+      args.push_back(Meet(op->args[i], tcn->args[i]));
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: should be invariant.

**Comment by MarisaKirisame** [comment]
Assuming you want to do subtyping, where every dimension is a subtype of Any (could be automatically casted to any),
There are two problems:

0: Union and Intersection (Join and Meet) cannot simply 'look all the way down' and only touch the Tensor.
They are structural change as well.
For example, you cannot Join a Ref (Tensor (1, 1) float) with Ref (Tensor (2, 1) float), and get Ref (Tensor (Any, 1) float). Ref (Tensor (1, 1) float) is not a subtype of (cannot be implicitly converted to) Ref (Tensor (Any, 1) float). Suppose you made a new Reference by implicitly converting, write (Tensor (2, 1) float) to it, and try to read using the old Reference, you will get a Tensor (2, 1) float, even though the type say the dimension must be (1, 1)!
Please read a bit about covariant/contravariant/invariant and understand how to do them.

1:
Subtyping is really, really hard. It make type inference hard. Is there any other way then subtyping? (for example, explicit casting?). I think basically any other option is better then subtyping.

**Comment by tqchen** [comment]
ping @icemelon9 @jroesch @MarisaKirisame please comment on the status of this PR, we can either mark as inactive and close it, or act on it to bring it in

**Comment by icemelon9** [comment]
close this for now

**Pull request closed without merging**